docs/notebooks/host-offloading.md

```diff
@@ -0,0 +1,351 @@
+
+- How to specify data distribution with sharding
+- How to control memory placement between host and device
+- How to manage data movement in jitted functions
+- How to control internal sharding within computations
+
+### NamedSharding and Memory Kinds
+
+{class}`~jax.sharding.NamedSharding` defines how data are distributed across devices. It includes:
+
+- Basic data distribution configuration
+- `memory_kind` parameter for specifying memory type (`device` or `pinned_host`)
+- By default, `memory_kind` is set to `device` memory
+- `with_memory_kind` method for creating new sharding with modified memory type
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: f-6sxUlqrlBn
+outputId: 79e7fbda-de0e-4951-9949-77039b2fae81
+---
+import jax
+import jax.numpy as jnp
+from jax.sharding import Mesh, NamedSharding, PartitionSpec as P
+import numpy as np
+
+# Create mesh
+# 1x1 mesh represents a single device with two named dimensions (x and y)
+mesh = Mesh(np.array(jax.devices()[0]).reshape(1,1), ('x','y'))
+
+# Device sharding - partitions data along x and y dimensions
+s_dev = NamedSharding(mesh, P('x', 'y'), memory_kind="device")
+
+# Host sharding - same partitioning but in pinned host memory
+s_host = s_dev.with_memory_kind('pinned_host')
+
+print(s_dev)   # Shows device memory sharding
+print(s_host)  # Shows pinned host memory sharding
+```
+
++++ {"id": "R_pB9465VoMP"}
+
+### Data Placement with device_put
+
+{func}`jax.device_put` is a function that explicitly transfers arrays to a specified memory location according to a sharding specification.
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+id: OJFnf7FGp6Lj
+outputId: a6c1fcdd-e49e-4017-c7aa-8be2e394c3a4
+---
+# Create a 4x8 array
+arr = jnp.arange(32.0).reshape(4, 8)
+
+# Move arrays to different memory locations based on sharding objects
+arr_host = jax.device_put(arr, s_host)  # Places in pinned host memory
+arr_dev = jax.device_put(arr, s_dev)    # Places in device memory
+
+# Verify memory locations
+print(arr_host.sharding.memory_kind)  # Output: pinned_host
+print(arr_dev.sharding.memory_kind)   # Output: device
+```
+
++++ {"id": "HHXvBpQKTMCR"}
+
+### Input/Output Sharding Controls
+
+Shardings determine how data are split across devices. JAX provides two key parameters for controlling data placement in jitted functions:
+1. `in_shardings`: controls how input arrays are partitioned when entering a jitted function
+2. `out_shardings`: controls how output arrays are partitioned when leaving a jitted function
+  - Can differ from input sharding
+  - Allows different memory kinds for outputs
+
+Example:
+
+```{code-cell} ipython3
+:id: ZXNj9NUeaIdX
+
+# Function with different input and output shardings
+def compute_function(x):
+  return x * 2
+
+compute_function = jax.jit(
```

Comment:
This example doesn't execute anything. Let's make sure all examples are executing something